In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns

In [2]:
df = pd.read_csv('../datasets/final_df.csv')
df.head()

total_sqft  bath   price  BHK  1st Block Jayanagar  1st Phase JP Nagar  \
0      1056.0   2.0   39.07  2.0                    0                   0   
1      1440.0   2.0   62.00  3.0                    0                   0   
2      1521.0   3.0   95.00  3.0                    0                   0   
3      1200.0   2.0   51.00  2.0                    0                   0   
4      2732.0   4.0  204.00  4.0                    0                   0   

   2nd Phase Judicial Layout  2nd Stage Nagarbhavi  5th Block Hbr Layout  \
0                          0                     0                     0   
1                          0                     0                     0   
2                          0                     0                     0   
3                          0                     0                     0   
4                          0                     0                     0   

   5th Phase JP Nagar  ...  Vijayanagar  Vishveshwarya Layout  \
0                   0  ...            0                     0   
1                   0  ...            0                     0   
2                   0  ...            0                     0   
3                   0  ...            0                     0   
4                   0  ...            0                     0   

   Vishwapriya Layout  Vittasandra  Whitefield  Yelachenahalli  Yelahanka  \
0                   0            0           0               0          0   
1                   0            0           0               0          0   
2                   0            0           0               0          0   
3                   0            0           0               0          0   
4                   0            0           0               0          0   

   Yelahanka New Town  Yelenahalli  Yeshwanthpur  
0                   0            0             0  
1                   0            0             0  
2                   0            0             0  
3                   0            0             0  
4                   0            0             0  

[5 rows x 243 columns]

In [3]:
X = df.drop(columns = ['price'])

In [4]:
y = df['price']

In [5]:
X

total_sqft  bath  BHK  1st Block Jayanagar  1st Phase JP Nagar  \
0         1056.0   2.0  2.0                    0                   0   
1         1440.0   2.0  3.0                    0                   0   
2         1521.0   3.0  3.0                    0                   0   
3         1200.0   2.0  2.0                    0                   0   
4         2732.0   4.0  4.0                    0                   0   
...          ...   ...  ...                  ...                 ...   
7950      1262.0   2.0  2.0                    0                   0   
7951      1345.0   2.0  3.0                    0                   0   
7952      1715.0   3.0  3.0                    0                   0   
7953      1141.0   2.0  2.0                    0                   0   
7954       550.0   1.0  1.0                    0                   0   

      2nd Phase Judicial Layout  2nd Stage Nagarbhavi  5th Block Hbr Layout  \
0                             0                     0                     0   
1                             0                     0                     0   
2                             0                     0                     0   
3                             0                     0                     0   
4                             0                     0                     0   
...                         ...                   ...                   ...   
7950                          0                     0                     0   
7951                          0                     0                     0   
7952                          0                     0                     0   
7953                          0                     0                     0   
7954                          0                     0                     0   

      5th Phase JP Nagar  6th Phase JP Nagar  ...  Vijayanagar  \
0                      0                   0  ...            0   
1                      0                   0  ...            0   
2                      0                   0  ...            0   
3                      0                   0  ...            0   
4                      0                   0  ...            0   
...                  ...                 ...  ...          ...   
7950                   0                   0  ...            0   
7951                   0                   0  ...            0   
7952                   0                   0  ...            0   
7953                   0                   0  ...            0   
7954                   0                   0  ...            0   

      Vishveshwarya Layout  Vishwapriya Layout  Vittasandra  Whitefield  \
0                        0                   0            0           0   
1                        0                   0            0           0   
2                        0                   0            0           0   
3                        0                   0            0           0   
4                        0                   0            0           0   
...                    ...                 ...          ...         ...   
7950                     0                   0            0           0   
7951                     0                   0            0           0   
7952                     0                   0            0           0   
7953                     0                   0            0           0   
7954                     0                   0            0           0   

      Yelachenahalli  Yelahanka  Yelahanka New Town  Yelenahalli  Yeshwanthpur  
0                  0          0                   0            0             0  
1                  0          0                   0            0             0  
2                  0          0                   0            0             0  
3                  0          0                   0            0             0  
4                  0          0                   0            0             0  
...         

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2)

**Ordinary Least Square Approach**

In [7]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()

In [8]:
lr.fit(X_train,y_train)

LinearRegression()

In [9]:
lr.predict(X_test)

array([ 94.01052909,  88.87504525, 161.62454501, ...,  31.85509178,
        97.10767931,  45.04762738])

In [10]:
lr.coef_

array([ 7.11019671e-02,  2.22024722e+00, -1.97163685e-01, -8.11833849e+00,
        1.79587006e+01, -3.55330552e+01, -4.20703049e+01, -5.23310321e+01,
       -2.14302982e+01,  7.04317673e+00, -1.84648221e+00, -2.51191800e+01,
       -2.37448169e+01, -2.09477774e+01, -2.75464790e+01, -2.01986669e+01,
       -7.13929882e+00, -2.01449655e+00, -1.41243588e+01, -2.29255571e+01,
       -3.07974785e+01, -1.91469789e+01, -2.59485581e+01, -1.34163795e+01,
       -1.71964807e+01, -1.29717292e+01, -2.19811530e+00,  4.07077908e-01,
       -1.16882215e+01, -2.84972269e+01, -7.83157425e+00,  8.68221761e-02,
       -1.14297188e+01,  1.57509153e+01, -1.23987346e+01, -4.68929183e+01,
       -5.50749821e+01, -1.30320895e+01, -6.60879654e+00,  5.72205003e+00,
       -1.34867062e+01,  3.83219648e+01,  9.08323434e+00, -2.62305950e+01,
       -2.91404582e+01, -3.42655689e+01, -8.82962696e+00,  2.10581670e+01,
       -2.50315450e+01, -4.65282332e+00, -5.14952165e+00,  1.77498629e+01,
       -2.31193700e+01, -

In [11]:
lr.intercept_

-8.670208537284026

In [12]:
y_test

2325     80.0
1575     90.0
5824    185.0
4686     38.0
6815     42.0
        ...  
6017    150.0
738     145.0
7646     35.0
1091    128.0
3919     48.0
Name: price, Length: 1591, dtype: float64

In [13]:
from sklearn.metrics import r2_score
y_pred = lr.predict(X_test)
print("R2Score = ",r2_score(y_test,y_pred))

R2Score =  0.8538583972145262


In [14]:
from sklearn.linear_model import Ridge,Lasso,ElasticNet

**Linear Regression**

In [15]:
R_0 = Ridge(alpha=0)
R_0.fit(X_train,y_train)
print("R2Score = ",r2_score(y_test,R_0.predict(X_test)))

R2Score =  0.8538583972145156


**For Ridge Regression, the maximum R2 score was achieved at alpha = 0.1**

In [16]:
R_10 = Ridge(alpha=0.1)
R_10.fit(X_train,y_train)
print("R2Score = ",r2_score(y_test,R_10.predict(X_test)))

R2Score =  0.854084730767709


**Lasso Regression**

In [17]:
L_0 = Lasso(alpha=0)
L_0.fit(X_train,y_train)
print("R2Score = ",r2_score(y_test,L_0.predict(X_test)))

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/base.py:1473: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  return fit_method(estimator, *args, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/linear_model/_coordinate_descent.py:697: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  model = cd_fast.enet_coordinate_descent(


R2Score =  0.8538583972145156


/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/linear_model/_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.695e+06, tolerance: 2.967e+03 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.linear_model.Ridge/RidgeCV instead.
  model = cd_fast.enet_coordinate_descent(


**For Lasso Regression, the maximum R2 score was achieved at alpha = 0.001**

In [18]:
L_10 = Lasso(alpha = 0.001)
L_10.fit(X_train,y_train)
print("R2Score = ",r2_score(y_test,L_10.predict(X_test)))

R2Score =  0.8540164256933038


**For ElasticNet Regression, the maximum R2 score was achieved at alpha = 0.0005 and l1_ratio = 0.8**

In [19]:
reg = ElasticNet(alpha=0.0005,l1_ratio=0.8)
reg.fit(X_train,y_train)
y_pred = reg.predict(X_test)
r2_score(y_test,y_pred)

0.8545466547310842

**The R2 score ranges from 0.853 to 0.854 in Linear,Ridge,Lasso and ElasticNet Regression.**

***So we will select ***Linear Regression*** as our model.***

In [20]:
def predict_price(location,sqft,bhk,bath):
    index_location = np.where(X.columns == location)[0][0]
    
    features = np.zeros(len(X.columns))
    features[0] = sqft
    features[1] = bath
    features[2] = bhk
    
    if index_location >= 1 :
        features[index_location] = 1
    
    return lr.predict([features])[0]

In [21]:
predict_price('Electronic City',1000,2,2)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


53.42928920085312

In [22]:
predict_price('Electronic City',1000,3,3)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


55.452372737196

In [23]:
predict_price('Electronic City',1000,4,4)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


57.47545627353887

In [24]:
predict_price('Electronic City',1000,1,1)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


51.40620566451024

In [25]:
predict_price('Electronic City',2000,2,3)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


126.75150350483281

In [26]:
predict_price('Koramangala',1000,2,3)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


84.11443022523198

In [28]:
import pickle
with open('model.pkl','wb') as file:
    pickle.dump(lr,file)